feat(registry): add scaleway_registry_namespace_beta resource

VoyTechnology opened the pull request:
Adds container registry and basic documentation associated with it.

QuentinBrosse commented on scaleway/provider.go:
```diff
@@ -212,6 +212,7 @@ func Provider() terraform.ResourceProvider {
 			"scaleway_security_group_rule":      resourceScalewaySecurityGroupRule(),
 			"scaleway_volume":                   resourceScalewayVolume(),
 			"scaleway_volume_attachment":        resourceScalewayVolumeAttachment(),
+			"scaleway_container_registry":       resourceScalewayContainerRegistry(),
```
We will name it `scaleway_registry_namespace_beta`.

We want to respect this pattern for all new resources: `scaleway_{productName}_{resourceName}[_beta]`. For every new resource, we prefer to release it with `_beta` suffix and remove it when we will be confident about the maturity of the resource.

Note that we have to rename:
- All functions: CRUD, helpers, tests, etc
- Documentation
- Files

Sorry for the inconvenience. 😕

VoyTechnology replied:
I hope I have covered all of the cases. If there are more please comment on them and I will change it.

QuentinBrosse commented on scaleway/resource_container_registry.go:
```diff
@@ -0,0 +1,139 @@
+		return err
+	}
+
+	_ = d.Set("name", ns.Name)
```
nit: you can remove `_ = `, just for consistency.

QuentinBrosse commented on scaleway/resource_container_registry.go:
```diff
@@ -0,0 +1,139 @@
+		if _, err := api.UpdateNamespace(&registry.UpdateNamespaceRequest{
+			Region:      region,
+			NamespaceID: id,
+			Description: &description,
```
```go
Description: scw.StringPtr(d.Get("description").(string)),
IsPublic: scw.BoolPtr(d.Get("is_public").(bool)),
```

QuentinBrosse commented on scaleway/resource_container_registry.go:
```diff
@@ -0,0 +1,139 @@
+	_ = d.Set("description", ns.Description)
+	_ = d.Set("organization_id", ns.OrganizationID)
+	_ = d.Set("is_public", ns.IsPublic)
+	_ = d.Set("endpoint", ns.Endpoint)
```
Add `d.Set("region", ns.Region)`

loicbourgois replied:
What about other fields documented at https://developers.scaleway.com/en/products/registry/api/#get-3cb8cf ?

loicbourgois replied:
If other read only fields are not Set, why set `endpoint` ?

QuentinBrosse replied:
We will expose only `name`, `description`, `organization_id`, `is_public`, `endpoint`, `region` and `id` for the moment. If a user has a use-case who requires another attribute we will talk about it at this moment.

QuentinBrosse commented on website/scaleway.erb:
```diff
@@ -124,6 +124,15 @@
           </ul>
         </li>
 
+        <li>
```
- Move it between `Baremetal Resources (Beta)` and `Kubernetes Resources (Beta)`.

QuentinBrosse commented on scaleway/resource_container_registry_test.go:
```diff
@@ -0,0 +1,98 @@
+		})
+
+		if err == nil {
+			return fmt.Errorf("Namespace (%s) still exists", rs.Primary.ID)
```
nit: `namespace`

QuentinBrosse commented on scaleway/resource_container_registry_test.go:
```diff
@@ -0,0 +1,98 @@
+						is_public = true
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
```
Add checks for `description` and `is_public`.

QuentinBrosse commented on website/docs/r/container_registry.markdown:
````diff
@@ -0,0 +1,54 @@
+
+```hcl
+resource "scaleway_container_registry" "main" {
+    region = "fr-par"
````
Remove `region` in basic example.

QuentinBrosse commented on website/docs/r/container_registry.markdown:
````diff
@@ -0,0 +1,54 @@
+}
+```
+
+## Arguments Reference
````
Can you move `region` and `organization_id` arguments after `is_public`? The majority of users will use the default ones.

AmFlint commented on website/docs/r/container_registry.markdown:
```diff
@@ -0,0 +1,54 @@
+
+- `description` (Optional) The description of the container registry namespace.
+
+- `is_public` (Defaults to `false`) Whether or not the registry should be visible to the public.
```
Whether or not the Registry Images stored in the namespace will be downloadable publicly (docker pull).

QuentinBrosse commented:
Hi @VoyTechnology and happy new year!

There is no activity in your PR since its creation on Nov 20, 2019. Are you available to finish it?
If you are not, we will close this PR et reopen another one to continue the work.

Thanks

VoyTechnology commented on scaleway/helper_container_registry.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+func containerRegistryWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```
Please see #336

VoyTechnology commented on scaleway/helper_container_registry.go:
```diff
@@ -0,0 +1,23 @@
+	return api, region, err
+}
+
+func containerRegistryWithRegionAndID(m interface{}, id string) (*registry.API, scw.Region, string, error) {
```
Please see #336

VoyTechnology commented:
I have renamed the resource as requested and rebased the code. Some of the comments are still valid, I am not sure how to continue on about them so I would appreciate extra details.

QuentinBrosse commented on scaleway/helper_container_registry.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+func registryNamespaceWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```
Rename the function (for consistency) and add a godoc:
```go
// containerRegistryAPIWithRegion returns a new container registry API and the region for a Create request`
func containerRegistryAPIWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```

QuentinBrosse commented on scaleway/helper_container_registry.go:
```diff
@@ -0,0 +1,23 @@
+	return api, region, err
+}
+
+func registryNamespaceWithRegionAndID(m interface{}, id string) (*registry.API, scw.Region, string, error) {
```
Same here (rename + godoc): `registryNamespaceAPIWithRegionAndID`

QuentinBrosse commented on scaleway/resource_registry_namespace_beta.go:
```diff
@@ -0,0 +1,139 @@
+		return err
+	}
+
+	name, ok := d.GetOk("name")
```
As the name is a unique identifier, we don't want to generate it and let the user explicitly specify it.

> Note that a namespace name is unique in a region. Thus, if another client already has created "test", you can't have it as a namespace.

You have marked the `name` attribute as required, it's fine, now you also have to remove the:
```go
if !ok {
	name = getRandomName("cr")
}
```

QuentinBrosse commented on scaleway/resource_registry_namespace_beta.go:
```diff
@@ -0,0 +1,139 @@
+		return err
+	}
+
+	_ = d.Set("name", ns.Name)
```
nit: for consistency, can you can remove `_ =` in all the following lines?

QuentinBrosse commented on scaleway/resource_registry_namespace_beta_test.go:
```diff
@@ -0,0 +1,98 @@
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayRegistryNamespaceBetaExists("scaleway_registry_namespace_beta.cr01"),
```
Remove empty line

VoyTechnology replied:
To which empty line are you referring to?

QuentinBrosse replied:
Sorry, it was a diff issue.

QuentinBrosse commented on scaleway/resource_registry_namespace_beta_test.go:
```diff
@@ -0,0 +1,98 @@
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayRegistryNamespaceBetaExists("scaleway_registry_namespace_beta.cr01"),
```
Remove empty line

VoyTechnology replied:
I'm sorry but I am not sure what line you are talking about.

QuentinBrosse replied:
Sorry, it was a diff issue.

QuentinBrosse commented on scaleway/resource_registry_namespace_beta_test.go:
```diff
@@ -0,0 +1,98 @@
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayRegistryNamespaceBetaExists("scaleway_registry_namespace_beta.cr01"),
+					testCheckResourceAttrUUID("scaleway_registry_namespace_beta.cr01", "id"),
```
Add checks for `description` and `is_public` attributes.

QuentinBrosse commented on website/docs/r/registry_namespace_beta.markdown:
```diff
@@ -0,0 +1,53 @@
+
+- `description` (Optional) The description of the container registry namespace.
+
+- `is_public` (Defaults to `false`) Whether or not the Registry Images stored in the namespace should be downloadable publicly (docker pull).
```
nit: "registry images" (in lowercase)

jerome-quere commented on website/scaleway.erb:
```diff
@@ -109,6 +109,15 @@
+          <a href="#">Container Registry Resources</a>
+          <ul class="nav nav-visible">
+            <li>
+              <a href="/docs/providers/scaleway/r/registry_namespace.html">scaleway_registry_namespace_beta</a>
```
Should be `/docs/providers/scaleway/r/registry_namespace_beta.html`

VoyTechnology commented:
The CI is failing because of
```
build github.com/terraform-providers/terraform-provider-scaleway: cannot load github.com/scaleway/scaleway-sdk-go/api/registry/v1: open /home/travis/gopath/src/github.com/terraform-providers/terraform-provider-scaleway/vendor/github.com/scaleway/scaleway-sdk-go/api/registry/v1: no such file or directory
```

I am not sure how does it relate to my PR, since the  `registry/v1` package is available in the version `1.0.0-beta5` used, and I have not changed anything in how the project is set up.

QuentinBrosse commented:
@VoyTechnology 
- To fix tests, can you try to get the last head version of the `scaleway-sdk-go` with `go get github.com/scaleway/scaleway-sdk-go@master` and/or rebase your branch to `master`?
- Can you also fix [this comment](https://github.com/terraform-providers/terraform-provider-scaleway/pull/330#discussion_r370512650), please?

Once this is done, your PR will be ready to merge. :)

VoyTechnology commented:
@QuentinBrosse The CI is now passing. Can you PTAL?

VoyTechnology commented:
Would you like me to squash all the commits before the merge?

QuentinBrosse commented:
Nop, I will do squash and merge with Github.
Thanks again for your great work @VoyTechnology! 🎉
